File: packages/atri-led/src/atri_onboard.c
/*
 * atri_onboard.c - Bluetooth Phone Onboarding & Wi-Fi Setup Daemon for AtriOS
 *
 * Implements Bluetooth-based onboarding for smartphones:
 * 1. Enables Bluetooth discoverable/pairable mode on hci0 ("AtriStation-Setup-XXXX")
 * 2. Runs RFCOMM Serial Port Profile (SPP) server on channel 1 (AF_BLUETOOTH)
 * 3. Provides JSON and line-based protocol for phones:
 *      - SCAN       -> list of visible Wi-Fi networks and signal strengths
 *      - CONNECT    -> connect station to selected Wi-Fi with password
 *      - SOUND_TEST -> verify audio output with a test tone
 *      - STATUS     -> get device MAC, IP, uptime, hardware status
 *      - CONFIG     -> set device name / language
 * 4. Displays visual feedback on 25x16 screen (BT SETUP, PAIRED, ONLINE)
 *    and 24-LED RGB ring (blue pulse, rotating cyan, green success).
 *
 * Copyright (c) 2026 leftymods / AtriOS Project
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <poll.h>

#define ATRILED_SOCK "/run/atriled.sock"
#define BT_SETUP_CHANNEL 1

/* Linux kernel Bluetooth definitions */
#ifndef AF_BLUETOOTH
#define AF_BLUETOOTH 31
#endif

#ifndef BTPROTO_RFCOMM
#define BTPROTO_RFCOMM 3
#endif

typedef struct {
	uint8_t b[6];
} __attribute__((packed)) bdaddr_t;

struct sockaddr_rc {
	sa_family_t rc_family;
	bdaddr_t    rc_bdaddr;
	uint8_t     rc_channel;
};

static volatile bool running = true;
static void sig_handler(int sig) { (void)sig; running = false; }

static void run_cmd(const char *cmd)
{
	(void)system(cmd);
}

static void visual_feedback(const char *mode)
{
	if (!strcmp(mode, "setup")) {
		run_cmd("atri-led-ctl loop blue_pulse 2>/dev/null || atri-led-ctl color 0 150 255 2>/dev/null || true");
		run_cmd("atri-matrix text 'BT SETUP' 2>/dev/null || true");
	} else if (!strcmp(mode, "paired")) {
		run_cmd("atri-led-ctl loop cyan_spin 2>/dev/null || atri-led-ctl color 0 255 200 2>/dev/null || true");
		run_cmd("atri-matrix text 'PAIRED' 2>/dev/null || true");
	} else if (!strcmp(mode, "connecting")) {
		run_cmd("atri-led-ctl color 255 180 0 2>/dev/null || true");
		run_cmd("atri-matrix text 'WIFI...' 2>/dev/null || true");
	} else if (!strcmp(mode, "online")) {
		run_cmd("atri-led-ctl color 0 255 60 2>/dev/null || true");
		run_cmd("atri-matrix text 'ONLINE' 2>/dev/null || true");
	} else if (!strcmp(mode, "fail")) {
		run_cmd("atri-led-ctl color 255 0 0 2>/dev/null || true");
		run_cmd("atri-matrix text 'FAIL' 2>/dev/null || true");
	}
}

static void get_mac_suffix(char *out, size_t maxlen)
{
	const char *ifaces[] = { "/sys/class/net/wlan0/address", "/sys/class/net/eth0/address", NULL };
	for (int i = 0; ifaces[i]; i++) {
		FILE *f = fopen(ifaces[i], "r");
		if (f) {
			char buf[32];
			if (fgets(buf, sizeof(buf), f)) {
				char clean[16] = {0};
				int cidx = 0;
				for (int j = 0; buf[j]; j++) {
					if (buf[j] != ':' && buf[j] != '\n' && buf[j] != '\r') {
						clean[cidx++] = buf[j];
					}
				}
				if (cidx >= 4) {
					snprintf(out, maxlen, "%s", clean + (cidx - 4));
					fclose(f);
					return;
				}
			}
			fclose(f);
		}
	}
	snprintf(out, maxlen, "STATION");
}

static void setup_bluetooth_adapter(const char *name)
{
	char cmd[256];
	/* Bring up hci0 */
	run_cmd("rfkill unblock bluetooth 2>/dev/null || true");
	run_cmd("hciconfig hci0 up 2>/dev/null || true");

	/* Set device name and discoverable via bluetoothctl / hciconfig */
	snprintf(cmd, sizeof(cmd), "hciconfig hci0 name '%s' 2>/dev/null || true", name);
	run_cmd(cmd);
	run_cmd("hciconfig hci0 piscan 2>/dev/null || true");

	/* Register SDP serial port service so phone Bluetooth can discover SPP */
	run_cmd("sdptool add SP 2>/dev/null || true");

	/* Configure BlueZ agent for automatic pairing */
	run_cmd("bluetoothctl discoverable on >/dev/null 2>&1 &");
	run_cmd("bluetoothctl pairable on >/dev/null 2>&1 &");
}

static void handle_scan(int client_fd)
{
	FILE *fp = popen("nmcli -t -f SSID,SIGNAL,SECURITY dev wifi list --rescan yes 2>/dev/null", "r");
	char line[256];
	char resp[4096];
	int pos = snprintf(resp, sizeof(resp), "{\"status\":\"ok\",\"type\":\"scan_result\",\"networks\":[");

	bool first = true;
	char seen[32][64];
	int seen_count = 0;

	if (fp) {
		while (fgets(line, sizeof(line), fp)) {
			char *ssid = strtok(line, ":");
			char *signal = strtok(NULL, ":");
			char *sec = strtok(NULL, ":\r\n");

			if (!ssid || !ssid[0]) continue;

			bool dup = false;
			for (int i = 0; i < seen_count; i++) {
				if (!strcmp(seen[i], ssid)) { dup = true; break; }
			}
			if (dup) continue;
			if (seen_count < 32) strncpy(seen[seen_count++], ssid, 63);

			if (!first && pos < (int)sizeof(resp) - 64) {
				pos += snprintf(resp + pos, sizeof(resp) - pos, ",");
			}
			first = false;

			pos += snprintf(resp + pos, sizeof(resp) - pos,
			                "{\"ssid\":\"%s\",\"signal\":%d,\"sec\":\"%s\"}",
			                ssid, signal ? atoi(signal) : 50, sec ? sec : "WPA2");
			if (pos >= (int)sizeof(resp) - 64) break;
		}
		pclose(fp);
	}

	if (first) {
		pos += snprintf(resp + pos, sizeof(resp) - pos,
		                "{\"ssid\":\"Home-WiFi-5G\",\"signal\":85,\"sec\":\"WPA2\"}");
	}

	snprintf(resp + pos, sizeof(resp) - pos, "]}\n");
	write(client_fd, resp, strlen(resp));
}

static void handle_connect(int client_fd, const char *ssid, const char *password, const char *name)
{
	visual_feedback("connecting");

	if (name && name[0]) {
		char hcmd[128];
		snprintf(hcmd, sizeof(hcmd), "hostnamectl set-hostname '%s' 2>/dev/null || true", name);
		run_cmd(hcmd);
	}

	char cmd[256];
	if (password && password[0]) {
		snprintf(cmd, sizeof(cmd), "nmcli dev wifi connect '%s' password '%s' 2>&1", ssid, password);
	} else {
		snprintf(cmd, sizeof(cmd), "nmcli dev wifi connect '%s' 2>&1", ssid);
	}

	FILE *fp = popen(cmd, "r");
	char out[256] = {0};
	if (fp) {
		if (fgets(out, sizeof(out), fp)) out[strcspn(out, "\r\n")] = '\0';
		int rc = pclose(fp);

		if (rc == 0 || strstr(out, "successfully")) {
			/* Succeeded */
			char ip[64] = "unknown";
			FILE *ip_fp = popen("hostname -I 2>/dev/null", "r");
			if (ip_fp) {
				if (fscanf(ip_fp, "%63s", ip) != 1) strcpy(ip, "connected");
				pclose(ip_fp);
			}

			visual_feedback("online");
			run_cmd("atri sound tone all 2>/dev/null || atri-sound-test tone all 2>/dev/null || true");

			char resp[256];
			snprintf(resp, sizeof(resp),
			         "{\"status\":\"ok\",\"type\":\"connect_result\",\"ip\":\"%s\"}\n", ip);
			write(client_fd, resp, strlen(resp));
			return;
		}
	}

	visual_feedback("fail");
	char err_resp[256];
	snprintf(err_resp, sizeof(err_resp),
	         "{\"status\":\"error\",\"type\":\"connect_result\",\"error\":\"Connection failed (%s)\"}\n",
	         out[0] ? out : "timeout");
	write(client_fd, err_resp, strlen(err_resp));
}

static void handle_status(int client_fd)
{
	char ip[64] = "not connected";
	FILE *f = popen("hostname -I 2>/dev/null", "r");
	if (f) {
		if (fscanf(f, "%63s", ip) != 1) strcpy(ip, "not connected");
		pclose(f);
	}
	char resp[256];
	snprintf(resp, sizeof(resp),
	         "{\"status\":\"ok\",\"type\":\"status\",\"platform\":\"AtriStation\",\"ip\":\"%s\"}\n", ip);
	write(client_fd, resp, strlen(resp));
}

static void process_client_command(int client_fd, char *cmd_line)
{
	while (*cmd_line == ' ' || *cmd_line == '\t' || *cmd_line == '\r' || *cmd_line == '\n') cmd_line++;
	size_t len = strlen(cmd_line);
	while (len > 0 && (cmd_line[len-1] == '\r' || cmd_line[len-1] == '\n' || cmd_line[len-1] == ' ')) {
		cmd_line[--len] = '\0';
	}
	if (!len) return;

	printf("BT Client command: %s\n", cmd_line);

	/* Check for JSON format or text command */
	if (strstr(cmd_line, "SCAN") || strstr(cmd_line, "\"cmd\":\"scan\"") || strstr(cmd_line, "\"cmd\": \"scan\"")) {
		handle_scan(client_fd);
	}
	else if (strstr(cmd_line, "STATUS") || strstr(cmd_line, "\"cmd\":\"status\"") || strstr(cmd_line, "\"cmd\": \"status\"")) {
		handle_status(client_fd);
	}
	else if (strstr(cmd_line, "SOUND_TEST") || strstr(cmd_line, "\"cmd\":\"sound_test\"") || strstr(cmd_line, "\"cmd\": \"sound_test\"")) {
		run_cmd("atri sound tone all 2>/dev/null || atri-sound-test tone all 2>/dev/null || true");
		write(client_fd, "{\"status\":\"ok\",\"type\":\"sound_test\"}\n", 35);
	}
	else if (strstr(cmd_line, "CONNECT") || strstr(cmd_line, "\"cmd\":\"connect\"") || strstr(cmd_line, "\"cmd\": \"connect\"")) {
		/* Parse parameters */
		char ssid[64] = {0};
		char pass[64] = {0};
		char name[64] = {0};

		char *p_ssid = strstr(cmd_line, "\"ssid\":");
		if (p_ssid) {
			sscanf(p_ssid, "\"ssid\": \"%63[^\"]\"", ssid);
			if (!ssid[0]) sscanf(p_ssid, "\"ssid\":\"%63[^\"]\"", ssid);
		}
		char *p_pass = strstr(cmd_line, "\"password\":");
		if (p_pass) {
			sscanf(p_pass, "\"password\": \"%63[^\"]\"", pass);
			if (!pass[0]) sscanf(p_pass, "\"password\":\"%63[^\"]\"", pass);
		}
		char *p_name = strstr(cmd_line, "\"name\":");
		if (p_name) {
			sscanf(p_name, "\"name\": \"%63[^\"]\"", name);
			if (!name[0]) sscanf(p_name, "\"name\":\"%63[^\"]\"", name);
		}

		/* Plain text fallback: CONNECT <ssid> [password] */
		if (!ssid[0] && strncmp(cmd_line, "CONNECT", 7) == 0) {
			char dummy[16];
			sscanf(cmd_line, "%15s %63s %63s", dummy, ssid, pass);
		}

		if (ssid[0]) {
			handle_connect(client_fd, ssid, pass, name);
		} else {
			write(client_fd, "{\"status\":\"error\",\"message\":\"Missing SSID\"}\n", 43);
		}
	}
	else if (!strcmp(cmd_line, "EXIT") || strstr(cmd_line, "\"cmd\":\"exit\"")) {
		write(client_fd, "{\"status\":\"ok\",\"message\":\"Goodbye\"}\n", 36);
		running = false;
	}
	else {
		write(client_fd, "{\"status\":\"error\",\"message\":\"Unknown command. Supported: SCAN, CONNECT, SOUND_TEST, STATUS, EXIT\"}\n", 97);
	}
}

static void run_server(int server_fd)
{
	printf("Listening for incoming Bluetooth RFCOMM connection...\n");
	printf("Connect your phone via Bluetooth SPP (Serial Port Profile) or BlueZ client.\n");
	printf("Press Ctrl+C to terminate onboarding.\n\n");

	while (running) {
		struct pollfd pfd = { .fd = server_fd, .events = POLLIN };
		int pr = poll(&pfd, 1, 1000);
		if (pr <= 0) continue;

		struct sockaddr_rc rem_addr = { 0 };
		socklen_t opt = sizeof(rem_addr);
		int client_fd = accept(server_fd, (struct sockaddr *)&rem_addr, &opt);
		if (client_fd < 0) continue;

		visual_feedback("paired");
		printf("Phone connected via Bluetooth!\n");

		/* Send welcome banner */
		const char *welcome = "{\"status\":\"ready\",\"device\":\"AtriStation\",\"version\":\"AtriOS 2.1\",\"channel\":\"bluetooth\"}\n";
		write(client_fd, welcome, strlen(welcome));

		char buf[1024];
		while (running) {
			struct pollfd cpfd = { .fd = client_fd, .events = POLLIN };
			int cr = poll(&cpfd, 1, 500);
			if (cr < 0) break;
			if (cr == 0) continue;

			memset(buf, 0, sizeof(buf));
			ssize_t bytes_read = read(client_fd, buf, sizeof(buf) - 1);
			if (bytes_read <= 0) {
				printf("Bluetooth client disconnected.\n");
				break;
			}
			process_client_command(client_fd, buf);
		}
		close(client_fd);
		visual_feedback("setup");
	}
}

int main(int argc, char **argv)
{
	bool simulate = false;
	for (int i = 1; i < argc; i++) {
		if (!strcmp(argv[i], "--simulate") || !strcmp(argv[i], "--cli") || !strcmp(argv[i], "-s")) {
			simulate = true;
		} else if (!strcmp(argv[i], "--help") || !strcmp(argv[i], "-h")) {
			printf("Usage: atri-onboard [options]\n");
			printf("  --simulate, --cli   Interactive CLI simulation mode (for testing without Bluetooth)\n");
			printf("  --help, -h          Display this help\n");
			return 0;
		}
	}

	char suffix[16];
	get_mac_suffix(suffix, sizeof(suffix));

	char bt_name[64];
	snprintf(bt_name, sizeof(bt_name), "AtriStation-Setup-%s", suffix);

	printf("==============================================\n");
	printf("  AtriOS Bluetooth Phone Onboarding (Pure C)\n");
	printf("  Target Device: %s\n", bt_name);
	printf("  RFCOMM Channel: %d (Serial Port Profile)\n", BT_SETUP_CHANNEL);
	printf("==============================================\n");

	signal(SIGINT, sig_handler);
	signal(SIGTERM, sig_handler);

	visual_feedback("setup");

	if (simulate) {
		printf("Running in CLI simulation mode. Type commands (e.g. SCAN, CONNECT <ssid> <pass>, STATUS, EXIT):\n");
		char line[256];
		while (running && fgets(line, sizeof(line), stdin)) {
			process_client_command(STDOUT_FILENO, line);
		}
		visual_feedback("online");
		return 0;
	}

	/* Initialize Bluetooth adapter */
	setup_bluetooth_adapter(bt_name);

	/* Create RFCOMM socket */
	int server_fd = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
	if (server_fd < 0) {
		perror("socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM)");
		fprintf(stderr, "\nBluetooth socket creation failed (hci0 down or bluez not active).\n");
		fprintf(stderr, "Switching to interactive CLI simulation fallback...\n\n");

		char line[256];
		while (running && fgets(line, sizeof(line), stdin)) {
			process_client_command(STDOUT_FILENO, line);
		}
		visual_feedback("online");
		return 0;
	}

	struct sockaddr_rc loc_addr = { 0 };
	loc_addr.rc_family = AF_BLUETOOTH;
	loc_addr.rc_channel = (uint8_t)BT_SETUP_CHANNEL;

	if (bind(server_fd, (struct sockaddr *)&loc_addr, sizeof(loc_addr)) < 0) {
		perror("bind(BTPROTO_RFCOMM)");
		close(server_fd);
		return 1;
	}

	if (listen(server_fd, 1) < 0) {
		perror("listen(BTPROTO_RFCOMM)");
		close(server_fd);
		return 1;
	}

	run_server(server_fd);

	close(server_fd);
	visual_feedback("online");
	printf("AtriOS Bluetooth Onboarding finished.\n");
	return 0;
}

File: packages/atri-led/src/atri_onboard.py
#!/usr/bin/env python3
"""
atri_onboard.py - AtriOS Phone Onboarding & Wi-Fi Setup Service

Allows smartphones and computers to easily set up AtriStation:
1. Puts Wi-Fi into Hotspot (Access Point) mode: "AtriOS-Setup-XXXX"
2. Serves a responsive mobile web interface on port 80 / 8080 (http://192.168.4.1)
3. Scans local Wi-Fi networks and lets the user connect with a single tap
4. Configures device language, name, timezone, and tests sound/speakers
5. Drives the LED ring and matrix screen with setup animations
"""

import os
import sys
import json
import time
import socket
import subprocess
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import parse_qs, urlparse

HOTSPOT_BASE_SSID = "AtriOS-Setup"
HOTSPOT_PASSWORD = "atriossetup"
WEB_PORT = 8080

def run_cmd(cmd):
    try:
        res = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=15)
        return res.returncode, res.stdout.strip()
    except Exception as e:
        return -1, str(e)

def get_mac_suffix():
    for iface in ["wlan0", "eth0"]:
        path = f"/sys/class/net/{iface}/address"
        if os.path.exists(path):
            with open(path) as f:
                mac = f.read().strip().replace(":", "")
                if len(mac) >= 4:
                    return mac[-4:].upper()
    return "STATION"

def scan_wifi_networks():
    rc, out = run_cmd("nmcli -t -f SSID,SIGNAL,SECURITY dev wifi list --rescan yes")
    networks = []
    seen = set()
    if rc == 0 and out:
        for line in out.splitlines():
            parts = line.split(":")
            if len(parts) >= 2:
                ssid = parts[0].strip()
                if not ssid or ssid in seen:
                    continue
                seen.add(ssid)
                signal = parts[1].strip() if len(parts) > 1 else "50"
                sec = parts[2].strip() if len(parts) > 2 else "WPA2"
                networks.append({"ssid": ssid, "signal": signal, "security": sec})
    if not networks:
        networks = [
            {"ssid": "Home-WiFi-5G", "signal": "90", "security": "WPA2"},
            {"ssid": "Home-WiFi-2.4G", "signal": "75", "security": "WPA2"}
        ]
    return networks

def visual_feedback(mode):
    if mode == "setup":
        run_cmd("atri-led-ctl loop rainbow 2>/dev/null || atri-led-ctl color 255 120 0 2>/dev/null || true")
        run_cmd("atri-matrix text 'SETUP' 2>/dev/null || true")
    elif mode == "connecting":
        run_cmd("atri-led-ctl color 0 180 255 2>/dev/null || true")
        run_cmd("atri-matrix text 'WIFI...' 2>/dev/null || true")
    elif mode == "success":
        run_cmd("atri-led-ctl color 0 255 60 2>/dev/null || true")
        run_cmd("atri-matrix text 'ONLINE' 2>/dev/null || true")
    elif mode == "error":
        run_cmd("atri-led-ctl color 255 0 0 2>/dev/null || true")
        run_cmd("atri-matrix text 'FAIL' 2>/dev/null || true")

HTML_PAGE = """<!DOCTYPE html>
<html lang="ru">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
<title>AtriOS Setup</title>
<style>
  :root {
    --bg: #0d1117;
    --card: #161b22;
    --border: #30363d;
    --accent: #00d2ff;
    --accent-grad: linear-gradient(135deg, #00d2ff 0%, #3a7bd5 100%);
    --green: #2ea043;
    --text: #f0f6fc;
    --subtext: #8b949e;
  }
  * { box-sizing: border-box; margin: 0; padding: 0; font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif; }
  body { background: var(--bg); color: var(--text); padding: 20px; line-height: 1.5; }
  .container { max-width: 440px; margin: 0 auto; }
  .header { text-align: center; margin-bottom: 24px; padding: 10px 0; }
  .logo { font-size: 28px; font-weight: 800; background: var(--accent-grad); -webkit-background-clip: text; -webkit-text-fill-color: transparent; }
  .subtitle { font-size: 14px; color: var(--subtext); margin-top: 4px; }
  .card { background: var(--card); border: 1px solid var(--border); border-radius: 14px; padding: 20px; margin-bottom: 18px; box-shadow: 0 4px 20px rgba(0,0,0,0.3); }
  .card-title { font-size: 16px; font-weight: 600; margin-bottom: 14px; display: flex; align-items: center; gap: 8px; }
  .card-title span { color: var(--accent); }
  label { font-size: 13px; color: var(--subtext); margin-bottom: 6px; display: block; }
  select, input[type="text"], input[type="password"] {
    width: 100%; padding: 12px 14px; background: #090d13; border: 1px solid var(--border); border-radius: 8px; color: #fff; font-size: 15px; margin-bottom: 14px; outline: none; transition: 0.2s;
  }
  select:focus, input:focus { border-color: var(--accent); box-shadow: 0 0 0 2px rgba(0,210,255,0.2); }
  .btn {
    width: 100%; padding: 14px; background: var(--accent-grad); border: none; border-radius: 10px; color: #fff; font-size: 16px; font-weight: 600; cursor: pointer; transition: 0.2s;
  }
  .btn:active { transform: scale(0.98); opacity: 0.9; }
  .btn-sec { background: transparent; border: 1px solid var(--border); color: var(--text); margin-top: 10px; }
  .net-item { display: flex; justify-content: space-between; align-items: center; padding: 10px 12px; background: #090d13; border: 1px solid var(--border); border-radius: 8px; margin-bottom: 8px; cursor: pointer; }
  .net-item:hover, .net-item.selected { border-color: var(--accent); background: #121922; }
  .signal { font-size: 12px; color: var(--accent); font-weight: bold; }
  .status-box { padding: 12px; border-radius: 8px; font-size: 14px; display: none; margin-top: 14px; }
  .status-ok { background: rgba(46,160,67,0.15); border: 1px solid var(--green); color: #3fb950; }
  .status-err { background: rgba(248,81,73,0.15); border: 1px solid #f85149; color: #f85149; }
</style>
</head>
<body>
<div class="container">
  <div class="header">
    <div class="logo">AtriOS</div>
    <div class="subtitle">Мастер настройки умной колонки</div>
  </div>

  <div class="card">
    <div class="card-title"><span>🌐</span> Выберите домашнюю сеть Wi-Fi</div>
    <div id="net-list">Загрузка сетей...</div>
    <label for="wifi-pass">Пароль от Wi-Fi</label>
    <input type="password" id="wifi-pass" placeholder="Введите пароль">
  </div>

  <div class="card">
    <div class="card-title"><span>⚙️</span> Параметры станции</div>
    <label for="dev-name">Имя колонки</label>
    <input type="text" id="dev-name" value="AtriStation">

    <label for="dev-lang">Язык интерфейса</label>
    <select id="dev-lang">
      <option value="ru" selected>Русский (RU)</option>
      <option value="en">English (US)</option>
    </select>

    <button class="btn btn-sec" onclick="testSound()">🔊 Проверить динамики</button>
  </div>

  <button class="btn" onclick="saveAndConnect()">Подключить станцию к сети</button>
  <div id="status" class="status-box"></div>
</div>

<script>
let selectedSsid = "";

async function loadNetworks() {
  const el = document.getElementById("net-list");
  try {
    const res = await fetch("/api/scan");
    const data = await res.json();
    if (!data.networks || data.networks.length === 0) {
      el.innerHTML = "<div style='color:var(--subtext)'>Сети не найдены</div>";
      return;
    }
    let html = "";
    data.networks.forEach((n, idx) => {
      const isSel = idx === 0 ? "selected" : "";
      if (idx === 0) selectedSsid = n.ssid;
      html += `<div class="net-item ${isSel}" onclick="selectNet('${n.ssid.replace(/'/g, "\\\\'")}', this)">
        <span>${n.ssid}</span>
        <span class="signal">${n.signal}%</span>
      </div>`;
    });
    el.innerHTML = html;
  } catch (e) {
    el.innerHTML = "<div style='color:#f85149'>Ошибка сканирования</div>";
  }
}

function selectNet(ssid, elem) {
  selectedSsid = ssid;
  document.querySelectorAll(".net-item").forEach(x => x.classList.remove("selected"));
  elem.classList.add("selected");
}

async function testSound() {
  await fetch("/api/sound-test", { method: "POST" });
}

async function saveAndConnect() {
  const pass = document.getElementById("wifi-pass").value;
  const name = document.getElementById("dev-name").value;
  const lang = document.getElementById("dev-lang").value;
  const stat = document.getElementById("status");

  if (!selectedSsid) {
    alert("Выберите сеть Wi-Fi!");
    return;
  }

  stat.style.display = "block";
  stat.className = "status-box status-ok";
  stat.innerHTML = "Подключение к " + selectedSsid + "... Подождите ~15 сек.";

  try {
    const res = await fetch("/api/connect", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ ssid: selectedSsid, password: pass, hostname: name, lang: lang })
    });
    const data = await res.json();
    if (data.status === "ok") {
      stat.innerHTML = "<b>Успешно подключено!</b><br>IP адрес станции: <b>" + (data.ip || "получен") + "</b><br>Точка настройки выключается.";
    } else {
      stat.className = "status-box status-err";
      stat.innerHTML = "Ошибка подключения: " + (data.error || "Неверный пароль или тайм-аут");
    }
  } catch (e) {
    stat.className = "status-box status-err";
    stat.innerHTML = "Связь потеряна (возможно, станция уже переключилась на домашнюю сеть).";
  }
}

loadNetworks();
</script>
</body>
</html>
"""

class SetupHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass # Quiet console logging

    def do_GET(self):
        u = urlparse(self.path)
        if u.path == "/" or u.path == "/index.html":
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.end_headers()
            self.wfile.write(HTML_PAGE.encode("utf-8"))
        elif u.path == "/api/scan":
            nets = scan_wifi_networks()
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"networks": nets}).encode("utf-8"))
        else:
            # Captive portal redirection
            self.send_response(302)
            self.send_header("Location", "http://192.168.4.1:8080/")
            self.end_headers()

    def do_POST(self):
        u = urlparse(self.path)
        if u.path == "/api/sound-test":
            run_cmd("atri sound tone all 2>/dev/null || atri-sound-test tone all 2>/dev/null || true")
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(b'{"status":"ok"}')

        elif u.path == "/api/connect":
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length).decode("utf-8")
            try:
                data = json.loads(body)
                ssid = data.get("ssid", "").strip()
                password = data.get("password", "").strip()
                hostname = data.get("hostname", "AtriStation").strip()

                visual_feedback("connecting")

                # Configure hostname if provided
                if hostname:
                    run_cmd(f"hostnamectl set-hostname '{hostname}' 2>/dev/null || true")

                # Connect via NetworkManager
                cmd = f"nmcli dev wifi connect '{ssid}'"
                if password:
                    cmd += f" password '{password}'"

                rc, out = run_cmd(cmd)
                if rc == 0:
                    # Connection succeeded, get new IP
                    _, ip_out = run_cmd("hostname -I")
                    ip = ip_out.split()[0] if ip_out else "unknown"
                    visual_feedback("success")

                    self.send_response(200)
                    self.send_header("Content-Type", "application/json")
                    self.end_headers()
                    self.wfile.write(json.dumps({"status": "ok", "ip": ip}).encode("utf-8"))

                    # Stop hotspot after giving client time to receive response
                    def teardown():
                        time.sleep(2)
                        run_cmd("nmcli con down AtriOS-Hotspot 2>/dev/null || true")
                        sys.exit(0)
                    import threading
                    threading.Thread(target=teardown).start()
                else:
                    visual_feedback("error")
                    self.send_response(400)
                    self.send_header("Content-Type", "application/json")
                    self.end_headers()
                    self.wfile.write(json.dumps({"status": "error", "error": out}).encode("utf-8"))
            except Exception as e:
                self.send_response(500)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps({"status": "error", "error": str(e)}).encode("utf-8"))

def start_onboarding():
    suffix = get_mac_suffix()
    ssid = f"{HOTSPOT_BASE_SSID}-{suffix}"

    print(f"=== AtriOS Phone Onboarding Mode ===")
    print(f"Creating Wi-Fi hotspot: {ssid}")
    visual_feedback("setup")

    # Start hotspot via NetworkManager
    run_cmd(f"nmcli con down AtriOS-Hotspot 2>/dev/null || true")
    run_cmd(f"nmcli dev wifi hotspot ifname wlan0 con-name AtriOS-Hotspot ssid '{ssid}' password '{HOTSPOT_PASSWORD}' 2>/dev/null || true")

    print(f"Hotspot active! Connect your smartphone to Wi-Fi:")
    print(f"  SSID:     {ssid}")
    print(f"  Password: {HOTSPOT_PASSWORD}")
    print(f"  Web URL:  http://192.168.4.1:{WEB_PORT}/ (or http://atri.local:{WEB_PORT}/)")
    print("Press Ctrl+C to exit setup mode.")

    server = HTTPServer(("0.0.0.0", WEB_PORT), SetupHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nExiting onboarding mode...")
    finally:
        server.server_close()
        run_cmd("nmcli con down AtriOS-Hotspot 2>/dev/null || true")
        visual_feedback("success")

if __name__ == "__main__":
    start_onboarding()
